source/common/protobuf/deterministic_hash.cc

```diff
@@ -0,0 +1,231 @@
+    break;
+  case FieldDescriptor::CPPTYPE_UINT64:
+    seed = hashScalarField<uint64_t>(reflection, message, field, seed);
+    break;
+  case FieldDescriptor::CPPTYPE_DOUBLE:
+    seed = hashScalarField<double>(reflection, message, field, seed);
+    break;
+  case FieldDescriptor::CPPTYPE_FLOAT:
+    seed = hashScalarField<float>(reflection, message, field, seed);
+    break;
+  case FieldDescriptor::CPPTYPE_BOOL:
+    seed = hashScalarField<bool>(reflection, message, field, seed);
+    break;
+  case FieldDescriptor::CPPTYPE_ENUM:
+    if (field.is_repeated()) {
+      const int c = reflection.FieldSize(message, &field);
+      for (int i = 0; i < c; i++) {
+        seed = HashUtil::xxHash64Value(reflection.GetRepeatedEnumValue(message, &field, i), seed);
+      }
+    } else {
+      seed = HashUtil::xxHash64Value(reflection.GetEnumValue(message, &field), seed);
+    }
+    break;
+  case FieldDescriptor::CPPTYPE_STRING:
+    if (field.is_repeated()) {
+      for (const std::string& str : reflection.GetRepeatedFieldRef<std::string>(message, &field)) {
+        seed = HashUtil::xxHash64(str, seed);
+      }
+    } else {
+      // Scratch may be used by GetStringReference if the field is not already a std::string.
+      std::string scratch;
+      seed = HashUtil::xxHash64(reflection.GetStringReference(message, &field, &scratch), seed);
+    }
+    break;
+  case FieldDescriptor::CPPTYPE_MESSAGE:
+    if (field.is_map()) {
+      seed = reflectionHashMapField(message, field, seed);
+    } else if (field.is_repeated()) {
+      for (const Protobuf::Message& submsg :
+           reflection.GetRepeatedFieldRef<Protobuf::Message>(message, &field)) {
+        seed = reflectionHashMessage(submsg, seed);
+      }
+    } else {
+      seed = reflectionHashMessage(reflection.GetMessage(message, &field), seed);
+    }
+    break;
+  }
+  return seed;
+}
+
+// Converts from type urls OR descriptor full names to descriptor full names.
+// Type urls are as used in envoy yaml config, e.g.
+// "type.googleapis.com/envoy.extensions.filters.udp.udp_proxy.v3.UdpProxyConfig"
+// becomes
+// "envoy.extensions.filters.udp.udp_proxy.v3.UdpProxyConfig"
+absl::string_view typeUrlToDescriptorFullName(absl::string_view url) {
+  const size_t pos = url.rfind('/');
+  if (pos != absl::string_view::npos) {
+    return url.substr(pos + 1);
+  }
+  return url;
+}
+
+std::unique_ptr<Protobuf::Message> unpackAnyForReflection(const ProtobufWkt::Any& any) {
+  const Protobuf::Descriptor* descriptor =
+      Protobuf::DescriptorPool::generated_pool()->FindMessageTypeByName(
+          typeUrlToDescriptorFullName(any.type_url()));
+  // If the type name refers to an unknown type, we treat it the same as other
+  // unknown fields - not including its contents in the hash.
+  if (descriptor == nullptr) {
+    return nullptr;
+  }
+  const Protobuf::Message* prototype =
+      Protobuf::MessageFactory::generated_factory()->GetPrototype(descriptor);
```

Comment:
The doc of `GetPrototype()` imply that the returned value will be null if the descriptor is unknown. I think you handled it above, but it may be better to add an ASSERT and a comment (just in case).

Reply:
Done, assuming ASSERT with a message functions as an adequate comment. :)